File: backend/app/services/local_auth_service.py
import secrets

from sqlalchemy.orm import Session

from app.core.passwords import hash_password, verify_password
from app.core.security import create_access_token
from app.models import LocalUser
from app.repositories.local_user_repository import LocalUserRepository


class LocalAuthService:
    def __init__(self, repo: LocalUserRepository):
        self._repo = repo

    def sign_up(self, db: Session, *, username: str, email: str | None, password: str) -> str:
        existing = self._repo.get_by_username(db, username=username)
        if existing is not None:
            raise ValueError("Username already exists")
        if email is not None:
            existing_email = self._repo.get_by_email(db, email=email)
            if existing_email is not None:
                raise ValueError("Email already exists")

        user = LocalUser(
            id=secrets.token_urlsafe(16),
            username=username,
            email=email,
            password_hash=hash_password(password),
        )
        self._repo.create(db, user=user)
        db.commit()

        return create_access_token(subject=user.id, provider="local", email=user.email)

    def sign_in(self, db: Session, *, username: str, password: str) -> str:
        user = self._repo.get_by_username(db, username=username)
        if user is None:
            raise ValueError("Invalid username or password")
        if not verify_password(password, user.password_hash):
            raise ValueError("Invalid username or password")

        return create_access_token(subject=user.id, provider="local", email=user.email)

File: dotnet_backend/Services/LocalAuthService.cs
using System.Security.Cryptography;
using dotnet_backend.Repositories;
using dotnet_backend.Core;

namespace dotnet_backend.Services;

public class LocalAuthService : ILocalAuthService
{
    private readonly ILocalUserRepository _repo;
    private readonly PasswordHasher _passwordHasher;
    private readonly JwtService _jwtService;

    public LocalAuthService(ILocalUserRepository repo, PasswordHasher passwordHasher, JwtService jwtService)
    {
        _repo = repo;
        _passwordHasher = passwordHasher;
        _jwtService = jwtService;
    }

    public string SignUp(string username, string? email, string password)
    {
        if (_repo.GetByUsername(username) != null)
            throw new ArgumentException("Username already exists");
        if (!string.IsNullOrEmpty(email) && _repo.GetByEmail(email) != null)
            throw new ArgumentException("Email already exists");

        var userId = Convert.ToBase64String(RandomNumberGenerator.GetBytes(16)).Replace("+", "-").Replace("/", "_").TrimEnd('=');
        var user = new Entities.LocalUser
        {
            Id = userId,
            Username = username,
            Email = email,
            PasswordHash = _passwordHasher.HashPassword(password)
        };
        _repo.Create(user);

        return _jwtService.CreateAccessToken(userId, "local", email);
    }

    public string SignIn(string username, string password)
    {
        var user = _repo.GetByUsername(username);
        if (user == null)
            throw new ArgumentException("User not found for the provided username.");
        if (!_passwordHasher.VerifyPassword(password, user.PasswordHash))
            throw new ArgumentException("Invalid password.");

        return _jwtService.CreateAccessToken(user.Id, "local", user.Email);
    }
}

